### ISSUE TITLE
User interaction required for evaluating SPO when using MFA

### ISSUE THREAD

**Issue by leitzler** [cisagov/ScubaGear#23]
# 🐛 Summary #

I'm trying to produce a report using a docker container as part of a headless workflow. As a part of the SharePoint rego `Get-SPOTenant` is used and it requires `Connect-SPOService` to auth before. According to [the documentation](https://learn.microsoft.com/en-us/powershell/module/sharepoint-online/connect-sposervice?view=sharepoint-ps) it require user interaction (or a global administrator username/password with MFA turned off).

The Graph API can use certificates to authenticate without user interaction, but as far as I can tell the baselines for Sharepoint & Onedrive aren't exposed via Graph API.

## To reproduce ##

1. Install dependencies
2. `Connect-MgGraph -ClientID <client id> -TentantId <tenant id> -CertificateThumbprint <thumbprint>`
3. `$InitDom = (Get-MgOrganization).VerifiedDomains | Where-Object {$_.isInitial}`
4. `$InitDomPrefix = $InitDom.Name.split(".")[0]`
5. `Connect-SPOService -Url "https://$($InitDomPrefix)-admin.sharepoint.com"`

## Expected behavior ##
To be able to evaluate baselines in a headless workflow.

## Any helpful log output or screenshots ##
```console
PS C:\> Connect-SPOService -Url "https://$($InitDomPrefix)-admin.sharepoint.com"
Connect-SPOService : No valid OAuth 2.0 authentication session exists
At line:1 char:1
+ Connect-SPOService -Url "https://$($InitDomPrefix)-admin.sharepoint.c ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Connect-SPOService], AuthenticationException
    + FullyQualifiedErrorId : Microsoft.Online.SharePoint.PowerShell.AuthenticationException,Microsoft.Online.SharePoint.PowerShell.ConnectSPOService
```

**Comment by gdasher** [comment]
Thanks for the feedback.  We are planning to "officially" add support for service principals as much as possible in a future release to support these kinds of headless use cases.  Its not in the release we're going to be releasing very soon, but hopefully relatively soon.  That said, the support will be incomplete as some of the underlying powershell modules use legacy APIs that don't support modern authentication or don't bubble up the necessary flexibility even if the underlying APIs have the required features.

When we add this feature, we'll do a deep dive to figure out exactly how much of the tool we can make work in a headlessly.  I think for SPO we may have to migrate to a different set of powershell modules or get a feature enhancement from Microsoft (c.f. https://techcommunity.microsoft.com/t5/sharepoint-developer/using-connect-sposervice-in-azure-runbooks-with-mfa-enabled/m-p/559941), but we will investigate.

**Comment by gdasher** [comment]
This issue has been resolved at head on the main branch by #96.  I'm going to go ahead and close this as a result.  @leitzler, please test it out if you're interested.  This functionality will be included in the 0.3.0 release when that comes out.

**Comment by leitzler** [comment]
Thank you @gdasher for looking into this!
I took a look at the latest commit and it seems like `Connect-PnP` is used instead of `Connect-SPOService` if I'm using a certificate?

And that causes several commands to fail:
```
WARNING: Error running Get-SPOTenant. No connection available. Use Connect-SPOService before running this CmdLet.
WARNING: Error running Get-SPOSite. No connection available. Use Connect-SPOService before running this CmdLet.
WARNING: Error running Get-SPOTenant. No connection available. Use Connect-SPOService before running this CmdLet.
WARNING: Error running Get-SPOTenantSyncClientRestriction. No connection available. Use Connect-SPOService before running this CmdLet.
```

If I understand it correctly those are required for onedrive/sharepoint evaulation?

**Comment by gdasher** [comment]
Yes, we use a different client library for SPO with certs.  We did test this pretty extensively though so it should work. Did you do a clean uninstall of the old modules before re-installing the new version?  utils/UninstallModules.ps1 and ./Setup.ps1 may help.

You'll see that the SPO provider uses different cmdlets when in pnp mode: https://github.com/cisagov/ScubaGear/blob/main/PowerShell/ScubaGear/Modules/Providers/ExportSharePointProvider.psm1

There are a few gaps, which is why we didn't do a wholesale migration to PnP.  Still working on that with Microsoft.

**Comment by leitzler** [comment]
Ah, thanks for the pointer! Turns out I missed setting `PnPFlag`, now things works as expected.
In my use case I'm calling the providers directly, partly to avoid the prior requirement of interaction.

As a side note, I noticed from #96 regarding permissions:
> SharePoint Permissions:
>
> Sites.Read.All or
> Site.FullControl.All
> // Was having trouble with just granting Sites.Read.All; please retest on other tenants to make sure

I can confirm that having `Sites.Read.All` wasn't enough for me, but adding `Site.FullControl.All` worked.
